packages/trie/src/proof/range.ts

```diff
@@ -382,6 +384,180 @@ async function hasRightElement(trie: Trie, key: Nibbles): Promise<boolean> {
   return false
 }
 
+export function isValueNode(node: TrieNode): boolean {
+  return (
+    node instanceof LeafNode ||
+    (node instanceof BranchNode && node.value() !== null && node.value()!.length > 0)
+  )
+}
+
+export function hasRightBranch(
+  node: BranchNode,
+  nibble: number
+): undefined | [number, EmbeddedNode] {
+  for (let i = nibble + 1; i < 16; i++) {
+    try {
+      const b = node.getBranch(i)
+      if (b !== null) {
+        return [i, b]
+      }
+    } catch (err) {
+      //
+    }
+  }
+  return undefined
+}
+
+export async function walkRight(this: Trie, key: Nibbles, hash: Uint8Array, childIdx: number) {
+  const rightWalk = this.walkTrieIterable(hash, [childIdx])
+  for await (const { node, currentKey } of rightWalk) {
+    if (nibblesCompare(key, [...currentKey, ...('_nibbles' in node ? node._nibbles : [])]) < 0) {
```

Comment:
walk iterator returns a stream of:
`{ node, currentKey }`

`currentKey` = nibbles collected on path to `node`.  Essentially the "full key" address of the parent of `node` ( + `branch_index` if parent is `branch` node)

`...('_nibbles' in node ? node._nibbles : [])]` 
is just a compact way of saying "if `node` is a leaf node, add the remaining nibbles, but if it's a branch node, don't even look for nibbles, just pretend it's an empty array".

In practice, though, since we're using hashed keys of equal length, this conditional should not be necessary, because all value nodes should be Leaf nodes.  

using nibbles compare here works because we're comparing to the actual key of the starting node that we found, which should always be the same length as the full key of the node.